node-packages/commons/src/tasks.js

```diff
@@ -211,10 +211,41 @@ async function createDeployTask(deployData: Object) {
 
   switch (project.activeSystemsDeploy) {
     case 'lagoon_openshiftBuildDeploy':
-      if (environments.project.productionEnvironment === branchName) {
+      // if (environments.project.productionEnvironment === branchName) {
+      //   logger.debug(
+      //     `projectName: ${projectName}, branchName: ${branchName}, production environment, no environment limits considered`,
+      //   );
+
+      // we want to limit production environments, without making it configurable currently
+      var productionEnvironmentsLimit = 2
+
+      // we want to make sure we can deploy the `production` env, and also any envs in the active/standby environment
+      // we may not even need the active/standby and just go to having `production` and `standbyProduction` only
+      if (
+        environments.project.productionEnvironment === branchName
+        || environments.project.standbyProductionEnvironment === branchName
+      ) {
+        // get a list of production environments
+        console.log(environments.project);
+        const prod_environments = environments.project.environments
+          .filter(e => e.environmentType === 'production')
+          .map(e => e.name);
         logger.debug(
-          `projectName: ${projectName}, branchName: ${branchName}, production environment, no environment limits considered`,
+          `projectName: ${projectName}, branchName: ${branchName}, existing environments are `,
+          prod_environments,
         );
+
+        if (prod_environments.length >= productionEnvironmentsLimit) {
+          if (prod_environments.find(i => i === branchName)) {
+            logger.debug(
+              `projectName: ${projectName}, branchName: ${branchName}, environment already exists, no environment limits considered`,
+            );
+          } else {
```

Comment:
This was when I was testing productionEnvironment to be able to specify branches as regex in the same way the branches field can. I'll clean this up.

Reply:
I've just looked over this again as this is meant to be similar enforcement to the development environment limit.

Part of A/S is to enforce a strict 2 production envs only. As currently, you can sort of trick lagoon into being able to deploy multiple production environments.

If I define `productionEnvironment` to be `brancha` and `standbyProductionEnvironment` to be `branchb`, and deploy both of these environments, they both will get created as production environments.

Both of these as part of `openshiftbuilddeploy` get set as `environmentType = 'production'`
https://github.com/amazeeio/lagoon/blob/6da12dfe7d0ae3d00d028cbf3011a46e7adb753d/services/openshiftbuilddeploy/src/index.js#L62-L68

If I then change `standbyProductionEnvironment` to be `branchc` then try and deploy it, it should not build it because brancha and branchb are already running as production `environmentType` in lagoon..

This should be the same as if I tried to deploy a development environment in that if I've deployed 5/5 already, then try to deploy another it will not do it.

Reply:
I think there's an edge case when an environment is marked as production but its name is not `productionEnvironment` or `standbyProductionEnvironment`. If that were the case then this new code branch gets skipped and it starts processing as if it were a dev environment instead. Who know what actually happens then.

I do agree though, that the only way to hit this edge case is for someone to manually change the environment type using the API. If you've tested all this an it's working then great, I'll leave final decision to you.